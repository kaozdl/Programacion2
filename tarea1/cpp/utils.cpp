/* 4580203 */
#include "../include/utils.h"

#include <stdio.h>

void imprimir_comp(const comp_t comp) {
  switch (comp) {
  case menor: {
    printf("<");
    /* En general, debe usarse break para evitar que siga procesando los otros
       casos. */
    break;
  }
  case igual: {
    printf("=");
    break;
  }
  case mayor: {
    printf(">");
    break;
  }
  }
} // imprimir_comp
