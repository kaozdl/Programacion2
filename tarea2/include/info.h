/*
  Módulo de definición de `info`.

  `info_t` es un puntero a una estructura (que debe definirse en
  info.cpp) cuyo contenido es un dato numérico (entero) y un dato de texto.

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
 */

#ifndef _INFO_T
#define _INFO_T

#include "utils.h"
#include "texto.h"

struct rep_info; // representación de `info`, definida en info.cpp
typedef rep_info *info_t;

/* Constructora, destructura y selectoras de info_t. */

/*
  Devuelve un info_t tal que su dato numérico es `n` y su dato de texto
  es `t`.
 */
info_t crear_info(const int n, const texto_t t);

/*
  Devuelve una copia de `i`.
  El elemento devuelto no comparte memoria con `i`.
 */
/*
info_t copiar_info(const info_t i);
Se incluirá en la próxima tarea.
*/

/*
  Libera la memoria reservada por `i`.
 */
void liberar_info(info_t &i);

/*
  Devuelve el dato numérico asociado a `i`.
 */
int numero_info(const info_t i);

/*
  Devuelve el dato de texto asociado a `i`.
 */
texto_t texto_info(const info_t i);

/*
  Devuelve `true` si `i` es válida o `false` en otro caso.
  `i` no es válida cuando su valor numérico es INT_MAX.
 */
bool es_valida_info(const info_t i);

/*
  Devuelve un texto_t formado por la concatenacion del dato numérico de `i`,
  el string "," y el dato de texto de `i` encerrados entre parentesis.
  El elemento devuelto no comparte memoria con `i`.
   Ejemplos:
   (4,a)
   (56,jojo)
 */
texto_t info_a_texto(const info_t i);

/*
  Lee texto desde la entrada estándar. Si tiene formato de `info_t` (ver
  info_a_texto) lo devuelve convertido en `info_t`. En otro caso devuelve una
  `info_t` con un  dato numérico = INT_MAX y dato de texto vacío.
  Precondición: hay texto para leer, que no incluye el símbolo ')' y tiene a
  lo sumo `max` caracteres.
 */
info_t leer_info(nat max);

#endif
