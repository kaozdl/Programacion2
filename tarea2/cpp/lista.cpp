/* 4580203 - 4660754 */
/*
  Módulo de implementación de 'lista'.

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
 */

#include "../include/lista.h"
#include "../include/info.h"

#include <stddef.h>

// nodo con punteros al anterior y al siguiente
struct nodo {
  info_t dato;
  nodo *anterior;   nodo *siguiente;
};

// lista implementada con cabecera con punteros al incio y al final
struct rep_lista {
  nodo *inicio;
  nodo *final;
};

lista crear_lista() {
  // se obtiene espacio para la cabecera
  // la lista vacía no es NULL
  lista segmento = new rep_lista;
  segmento->inicio = segmento->final = NULL;
  return segmento;
}

void insertar_antes(const info_t i, const localizador loc, lista &lst) {
  nodo *nuevo = new nodo;
  nuevo->dato = i;
  if (es_vacia_lista(lst)) {
    // nuevo es el único nodo, por lo que es el inicio y el final
    lst->inicio = lst->final = nuevo;
    nuevo->anterior = nuevo->siguiente = NULL;
  } else {
    nuevo->siguiente = loc;
    nuevo->anterior = loc->anterior;
    loc->anterior = nuevo;

    if (loc == lst->inicio)
      lst->inicio = nuevo;
    else
      nuevo->anterior->siguiente = nuevo;
  }
}

lista segmento_lista(const localizador desde, const localizador hasta,
                     const lista lst) {
  lista segmento = crear_lista();
  if (!es_vacia_lista(lst)) {
    localizador loc = desde;
    while (loc != hasta->siguiente) {
      // ERROR: compartiría memoria
      // info_t info = loc->dato;

      // ERROR: compartiría memoria
      // info_t info = crear_info(numero_info(loc->dato),
      // texto_info(loc->dato));

      info_t info = crear_info(numero_info(loc->dato),
                               copiar_texto(texto_info(loc->dato)));
      insertar_despues(info, final_lista(segmento), segmento);
      loc = siguiente(loc, lst);
    }
  }
  return segmento;
}

bool es_localizador_lista(const localizador loc) { return loc != NULL; }

bool localizador_pertenece_a_lista(const localizador loc, const lista lst) {
  bool segmento = false;
  localizador cursor = lst->inicio;
  while ((!segmento) && (es_localizador_lista(cursor))) {
    if (cursor == loc)
      segmento = true;     else
      cursor = cursor->siguiente;
  }
  return segmento;
}

localizador siguiente_clave(const int clave, const localizador loc,
                            const lista lst) {
  localizador segmento;
  if (es_vacia_lista(lst))
    segmento = NULL;
  else {
    segmento = loc;
    bool encontrado = false;
    while ((!encontrado) && (segmento != NULL)) {
      if (numero_info(segmento->dato) == clave)
        encontrado = true;
      else
        segmento = segmento->siguiente;
    }
  }
  return segmento;
}

void mover_antes(localizador loc1, localizador loc2, lista &lst) {
  // En esta versión no se solicita memoria para crear nodos, sino que
  // se mueve todo el nodo que es accedido con `loc2`.
  if (loc1 != loc2) {
    // La creación de una lista requiere espacio para la cabecera que habrá
    // que liberar al final.
    lista sgm = separar_segmento(loc2, loc2, lst);

    // El segemento (de un elemento) se inserta antes de `loc1`.
    // Es lo que se haría si existiera la funcióm insertar_segmento_antes
    // excepto en que acá no es necesario contemplar el caso en que la lista
    // es vacía ya que se sabe que loc1 != loc2
    sgm->inicio->anterior = loc1->anterior;
    sgm->final->siguiente = loc1;
    if (es_inicio_lista(loc1, lst)) {
      lst->inicio = sgm->inicio;
    } else {
      loc1->anterior->siguiente = sgm->inicio;
    }
    loc1->anterior = sgm->final;

    // Se libera la memoria de la cabecera de la lista, ya vacía.
    delete sgm;
  }
}


void insertar_despues(const info_t i, const localizador loc, lista &lst) {
  nodo *nuevo = new nodo;
  nuevo->dato = i;
  nuevo->siguiente = NULL;
  nuevo->anterior = NULL;
  if (es_vacia_lista(lst)) {
	   lst->inicio = lst->final = nuevo;
      	}
	else{
    if (loc != lst->final){
      loc->siguiente->anterior = nuevo;
      nuevo->siguiente = loc->siguiente;
      nuevo->anterior = loc;
      loc->siguiente = nuevo;
    }
    else {
      loc->siguiente = nuevo;
      nuevo->anterior = loc;
      lst->final = nuevo;
    }
  }
}

void insertar_segmento_despues(lista &sgm, const localizador loc, lista &lst) {
  if (es_vacia_lista (lst) && (!es_vacia_lista(sgm))){
    lst->inicio = sgm->inicio;
    lst->final = sgm->final;
  }
  else if ((loc->siguiente != NULL) && (!es_vacia_lista(sgm))) {
    localizador resto = loc->siguiente;
    loc->siguiente = sgm->inicio;
    sgm->final->siguiente = resto;
    resto->anterior = sgm->final;
    sgm->inicio->anterior = loc;
  }
      else if (!es_vacia_lista(sgm)) {
        loc->siguiente = sgm->inicio;
        sgm->inicio->anterior = loc;
        lst->final = sgm->final;

  }
  sgm->inicio = NULL;
  sgm->final = NULL;
}

lista separar_segmento(localizador desde, localizador hasta, lista &lst) {
    lista segmento = crear_lista();
    if (lst->inicio!=NULL){
        segmento->inicio = desde;
        segmento->final = hasta;
        if (segmento->final->siguiente!=NULL)
            segmento->final->siguiente->anterior = desde->anterior;
        else
            lst->final=desde->anterior;
        if (segmento->inicio->anterior !=NULL)
            segmento->inicio->anterior->siguiente = hasta->siguiente;
        else
            lst->inicio=hasta->siguiente;
        desde->anterior=NULL;
        hasta->siguiente=NULL;
    }
    return segmento;
}

void remover_de_lista(localizador &loc, lista &lst) {
if (!es_vacia_lista(lst)){
    if (es_localizador_lista(loc->siguiente))
        loc->siguiente->anterior= loc->anterior;
    else
        lst->final=loc->anterior;
    if (es_localizador_lista(loc->anterior))
            loc->anterior->siguiente = loc->siguiente;
    else
        lst->inicio = loc->siguiente;
      delete loc;
    }
}

void liberar_lista(lista &lst) {
  localizador loc = lst->final;
  while (loc!=NULL){
    localizador anterior = loc->anterior;
    delete loc;
    loc = anterior;
  }
  delete lst;
  lst=NULL;
}

bool es_vacia_lista(const lista lst) {
  return (lst->inicio==NULL);
}

bool es_inicio_lista(const localizador loc, const lista lst) {
  return (loc==lst->inicio);
}

bool es_final_lista(const localizador loc, const lista lst) {
  return (loc==lst->final);
}

localizador inicio_lista(const lista lst) {
  return lst->inicio;
}

localizador final_lista(const lista lst) {
  return lst->final;
}

localizador siguiente(const localizador loc, const lista lst) {
  return loc->siguiente;
}

localizador anterior(const localizador loc, const lista lst) {
  return loc->anterior;
}

bool precede_en_lista(const localizador l1, const localizador l2,const lista lst) {
  localizador localizador1 = l2;
  while (es_localizador_lista(localizador1) && localizador1!=l1)
    localizador1=anterior(localizador1,lst);
  return es_localizador_lista(localizador1);
}

localizador anterior_clave(const int clave, const localizador loc,const lista lst) {
  localizador localizador1 = loc;
  if(es_localizador_lista(localizador1)){
    while (es_localizador_lista(localizador1) && numero_info(info_lista(localizador1,lst))!=clave)
      localizador1 =anterior(localizador1,lst);
      return localizador1;
    }
  else return NULL;
}

info_t info_lista(const localizador loc, const lista lst) {
  return loc->dato;
}

void cambiar_en_lista(const info_t i, const localizador loc, lista &lst) {
  loc->dato=i;
}

void intercambiar(const localizador loc1, const localizador loc2, lista &lst) {
  info_t info_auxiliar = loc1->dato;
  loc1->dato = loc2->dato;
  loc2->dato = info_auxiliar;
}
