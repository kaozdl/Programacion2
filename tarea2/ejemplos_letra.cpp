/*
  Módulo principal ejemplos_letra.

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
*/

#include "include/utils.h"
#include "include/texto.h"
#include "include/info.h"
#include "include/lista.h"
#include "include/uso_lista.h"

#include <stddef.h>
#include <stdio.h>

int main() {
  lista lst, lst2, sgm;
  info_t info;
  localizador loc, loc2;

  // insertar-antes
  printf("insertar-antes. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();

  info = crear_info(3, cadena_a_texto("t"));
  loc = inicio_lista(lst);
  insertar_antes(info, loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(8, cadena_a_texto("o"));
  insertar_despues(info, final_lista(lst), lst);

  info = crear_info(3, cadena_a_texto("t"));
  loc = inicio_lista(lst);
  insertar_antes(info, loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // tercer ejemplo
  printf("3:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(8, cadena_a_texto("o"));
  insertar_despues(info, final_lista(lst), lst);

  info = crear_info(3, cadena_a_texto("t"));
  loc = anterior(final_lista(lst), lst);
  insertar_antes(info, loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // remover_de_lista
  printf("remover_de_lista. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();

  loc = inicio_lista(lst);
  remover_de_lista(loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(8, cadena_a_texto("o"));
  insertar_despues(info, final_lista(lst), lst);

  loc = inicio_lista(lst);
  remover_de_lista(loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // tercer ejemplo
  printf("3:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(8, cadena_a_texto("o"));
  insertar_despues(info, final_lista(lst), lst);

  loc = final_lista(lst);
  remover_de_lista(loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // cuarto ejemplo
  printf("4:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(8, cadena_a_texto("o"));
  insertar_despues(info, final_lista(lst), lst);

  loc = anterior(final_lista(lst), lst);
  remover_de_lista(loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // intercambiar
  printf("intercambiar. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  loc = siguiente(inicio_lista(lst), lst);
  loc2 = anterior(final_lista(lst), lst);
  intercambiar(loc, loc2, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // siguiente_clave
  printf("siguiente_clave. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();
  loc = inicio_lista(lst);
  loc = siguiente_clave(8, loc, lst);
  if (es_localizador_lista(loc)) {
    texto_t t = info_a_texto(info_lista(loc, lst));
    escribir_texto(t);
    liberar_texto(t);
  }
  escribir_nueva_linea();
  liberar_lista(lst);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);

  loc = siguiente(inicio_lista(lst), lst);
  loc = siguiente_clave(3, loc, lst);
  if (es_localizador_lista(loc)) {
    texto_t t = info_a_texto(info_lista(loc, lst));
    escribir_texto(t);
    liberar_texto(t);
  }
  escribir_nueva_linea();
  liberar_lista(lst);

  // tercer ejemplo
  printf("3:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(4, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(11, cadena_a_texto("o"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);

  loc = siguiente(inicio_lista(lst), lst);
  loc = siguiente_clave(11, loc, lst);
  if (es_localizador_lista(loc)) {
    texto_t t = info_a_texto(info_lista(loc, lst));
    escribir_texto(t);
    liberar_texto(t);
  }
  escribir_nueva_linea();
  liberar_lista(lst);

  // separar_segmento
  printf("separar_segmento. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  loc = inicio_lista(lst);
  loc2 = anterior(final_lista(lst), lst);
  sgm = separar_segmento(loc, loc2, lst);

  imprimir_lista(sgm);
  imprimir_lista(lst);
  liberar_lista(lst);
  liberar_lista(sgm);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  loc = siguiente(siguiente(inicio_lista(lst), lst), lst);
  loc2 = final_lista(lst);
  sgm = separar_segmento(loc, loc2, lst);

  imprimir_lista(sgm);
  imprimir_lista(lst);
  liberar_lista(lst);
  liberar_lista(sgm);

  // tercer ejemplo
  printf("3:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  loc = siguiente(inicio_lista(lst), lst);
  loc2 = anterior(final_lista(lst), lst);
  sgm = separar_segmento(loc, loc2, lst);

  imprimir_lista(sgm);
  imprimir_lista(lst);
  liberar_lista(lst);
  liberar_lista(sgm);

  // reversa
  printf("reversa. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();

  sgm = reversa(lst);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(sgm);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  sgm = reversa(lst);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(sgm);

  // retroceder
  printf("retroceder. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  loc = inicio_lista(lst);
  retroceder(loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  loc = final_lista(lst);
  retroceder(loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // tercer ejemplo
  printf("3:");
  lst = crear_lista();
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(9, cadena_a_texto("n"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  loc = anterior(final_lista(lst), lst);
  retroceder(loc, lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // unificar
  printf("unificar. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  unificar(lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("u"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("t"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(5, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(6, cadena_a_texto("s"));
  insertar_despues(info, final_lista(lst), lst);

  unificar(lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // tercer ejemplo
  printf("3:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("e"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(1, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(1, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(1, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(1, cadena_a_texto("a"));
  insertar_despues(info, final_lista(lst), lst);

  unificar(lst);

  imprimir_lista(lst);
  liberar_lista(lst);

  // mezcla
  printf("mezcla. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();
  lst2 = crear_lista();

  sgm = mezcla(lst, lst2);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(lst2);
  liberar_lista(sgm);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  lst2 = crear_lista();
  info = crear_info(1, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst2), lst2);
  info = crear_info(2, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst2), lst2);
  info = crear_info(7, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst2), lst2);

  sgm = mezcla(lst, lst2);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(lst2);
  liberar_lista(sgm);

  // tercer ejemplo
  printf("3:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("a"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("j"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("r"));
  insertar_despues(info, final_lista(lst), lst);
  lst2 = crear_lista();
  info = crear_info(1, cadena_a_texto("d"));
  insertar_despues(info, final_lista(lst2), lst2);
  info = crear_info(2, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst2), lst2);
  info = crear_info(7, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst2), lst2);

  sgm = mezcla(lst, lst2);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(lst2);
  liberar_lista(sgm);

  // filtrado
  printf("filtrado. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();
  info = crear_info(7, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(1, cadena_a_texto("a"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);

  sgm = filtrado(2, menor, lst);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(sgm);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(7, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(1, cadena_a_texto("a"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);

  sgm = filtrado(8, igual, lst);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(sgm);

  // tercer ejemplo
  printf("3:");
  lst = crear_lista();
  info = crear_info(7, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(1, cadena_a_texto("a"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);

  sgm = filtrado(1, mayor, lst);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(sgm);

  // sublista
  printf("sublista. \n");
  // primer ejemplo
  printf("1:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("a"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(7, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);

  sgm = sublista(1, 2, lst);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(sgm);

  // segundo ejemplo
  printf("2:");
  lst = crear_lista();
  info = crear_info(1, cadena_a_texto("a"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(2, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("a"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(3, cadena_a_texto("b"));
  insertar_despues(info, final_lista(lst), lst);
  info = crear_info(7, cadena_a_texto("c"));
  insertar_despues(info, final_lista(lst), lst);

  sgm = sublista(3, 7, lst);

  imprimir_lista(sgm);
  liberar_lista(lst);
  liberar_lista(sgm);

  return 0;
}
