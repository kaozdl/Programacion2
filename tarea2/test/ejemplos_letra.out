insertar-antes. 
1:(3,t)
2:(3,t)(1,u)(5,c)(2,d)(8,o)
3:(1,u)(5,c)(3,t)(2,d)(8,o)
remover_de_lista. 
1:
2:(5,c)(2,d)(8,o)
3:(1,u)(5,c)(2,d)
4:(1,u)(5,c)(8,o)
intercambiar. 
1:(1,u)(9,n)(5,c)(3,t)(6,s)
siguiente_clave. 
1:
2:
3:(11,o)
separar_segmento. 
1:(1,u)(3,t)(5,c)(9,n)
(6,s)
2:(5,c)(9,n)(6,s)
(1,u)(3,t)
3:(3,t)(5,c)(9,n)
(1,u)(6,s)
reversa. 
1:
2:(6,s)(9,n)(5,c)(3,t)(1,u)
retroceder. 
1:(1,u)(3,t)(5,c)(9,n)(6,s)
2:(1,u)(3,t)(5,c)(6,s)(9,n)
3:(1,u)(3,t)(5,c)(9,n)(6,s)
unificar. 
1:(1,u)(3,t)(5,c)(6,s)
2:(1,u)(3,t)(5,c)(6,s)
3:(1,e)
mezcla. 
1:
2:(1,d)(2,b)(7,c)
3:(1,a)(1,d)(2,j)(2,b)(3,r)(7,c)
filtrado. 
1:(1,a)
2:
3:(7,c)(2,b)
sublista. 
1:(1,a)(2,b)
2:(3,a)(3,c)(3,b)(7,c)
